codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/EndpointResolverGenerator.kt

```diff
@@ -0,0 +1,281 @@
+                """
+                /// The default endpoint resolver
+                pub struct DefaultResolver {
+                    #{custom_fields:W}
+                }
+
+                impl DefaultResolver {
+                    /// Create a new endpoint resolver with default settings
+                    pub fn new() -> Self { 
+                        Self { #{custom_fields_init:W} }
+                    }
+                }
+
+                impl #{endpoint}::ResolveEndpoint<#{Params}> for DefaultResolver {
+
+                    fn resolve_endpoint(&self, params: &Params) -> #{endpoint}::Result {
+                        let mut diagnostic_collector = #{DiagnosticCollector}::new();
+                        #{resolver_fn}(params, &mut diagnostic_collector, #{additional_args})
+                            .map_err(|err|err.with_source(diagnostic_collector.take_last_error()))
+                    }
+                }
+                """,
+                "custom_fields" to fnsUsed.mapNotNull { it.structField() }.join(","),
+                "custom_fields_init" to fnsUsed.mapNotNull { it.structFieldInit() }.join(","),
+                "Params" to EndpointParamsGenerator(endpointRuleSet.parameters).paramsStruct(),
+                "additional_args" to fnsUsed.mapNotNull { it.additionalArgsInvocation("self") }.join(","),
+                "resolver_fn" to resolverFn(endpointRuleSet, registry, fnsUsed),
+                *codegenScope,
+            )
+        }
+    }
+
+    private fun resolverFn(
+        endpointRuleSet: EndpointRuleSet,
+        registry: FunctionRegistry,
+        fnsUsed: List<CustomRuntimeFunction>,
+    ): RuntimeType {
+        return RuntimeType.forInlineFun("resolve_endpoint", EndpointsImpl) {
+            rustTemplate(
+                """
+                pub(super) fn resolve_endpoint($ParamsName: &#{Params}, $DiagnosticCollector: &mut #{DiagnosticCollector}, #{additional_args}) -> #{endpoint}::Result {
+                 #{body:W}
+                }
+
+                """,
+                *codegenScope,
+                "Params" to EndpointParamsGenerator(endpointRuleSet.parameters).paramsStruct(),
+                "additional_args" to fnsUsed.mapNotNull { it.additionalArgsSignature() }.join(","),
+                "body" to resolverFnBody(endpointRuleSet, registry),
+            )
+        }
+    }
+
+    private fun resolverFnBody(endpointRuleSet: EndpointRuleSet, registry: FunctionRegistry) = writable {
+        val scope = Scope.empty()
+        endpointRuleSet.parameters.toList().forEach {
+            Attribute.AllowUnused.render(this)
+            rust("let ${it.memberName()} = &$ParamsName.${it.memberName()};")
+        }
+        generateRulesList(endpointRuleSet.rules, scope)(this)
+    }
+
+    private fun generateRulesList(rules: List<Rule>, scope: Scope) = writable {
+        rules.forEach { rule ->
+            rule.documentation.orNull()?.also { docs(it, newlinePrefix = "//") }
```

Comment:
We specify the newline enough that it would prob be worth it to alias the `docs` fn and have one for documentation and one for comments. Maybe one for for inner-line docs too?

https://doc.rust-lang.org/reference/comments.html